packages/gatsby/src/schema/types/pagination.ts

```diff
@@ -0,0 +1,134 @@
+import {
+  SchemaComposer,
+  ObjectTypeComposer,
+  InputTypeComposer,
+  InterfaceTypeComposer,
+} from "graphql-compose"
+
+import { getFieldsEnum } from "./sort"
+import { addDerivedType } from "./derived-types"
+import { distinct, group } from "../resolvers"
+
+const getPageInfo = ({
+  schemaComposer,
+}: {
+  schemaComposer: SchemaComposer<any>
+}): ObjectTypeComposer =>
+  schemaComposer.getOrCreateOTC(`PageInfo`, tc => {
+    tc.addFields({
+      currentPage: `Int!`,
+      hasPreviousPage: `Boolean!`,
+      hasNextPage: `Boolean!`,
+      itemCount: `Int!`,
+      pageCount: `Int!`,
+      perPage: `Int`,
+    })
+  })
+
+const getEdge = ({
+  schemaComposer,
+  typeComposer,
+}: {
+  schemaComposer: SchemaComposer<any>
+  typeComposer: ObjectTypeComposer | InterfaceTypeComposer
+}): ObjectTypeComposer => {
+  const typeName: string = typeComposer.getTypeName() + `Edge`
+  addDerivedType({ typeComposer, derivedTypeName: typeName })
+  return schemaComposer.getOrCreateOTC(typeName, tc => {
+    tc.addFields({
+      next: typeComposer,
+      node: typeComposer.getTypeNonNull(),
+      previous: typeComposer,
+    })
+  })
+}
+
+const createPagination = ({
+  schemaComposer,
+  typeComposer,
+  fields,
+  typeName,
+}: {
+  schemaComposer: SchemaComposer<any>
+  typeComposer: ObjectTypeComposer | InterfaceTypeComposer
+  // TODO: Define the interface for the 'fields' data structure
+  fields: Record<string, any>
+  typeName: string
+}): ObjectTypeComposer => {
```

Comment:
You could type it something like this:

```typescript
const createPagination = <TContext = any, TSource = any>({
  schemaComposer,
  typeComposer,
  fields,
  typeName,
}: {
  schemaComposer: SchemaComposer<TContext>
  typeComposer: ObjectTypeComposer | InterfaceTypeComposer
  fields: ComposeFieldConfigMap<TSource, TContext>
  typeName: string
}): ObjectTypeComposer => {
```

Reply:
Great!!
@ascorbic, I extended your suggestion to the other methods.